greedy_algorithms/gale_shapley.cpp

```diff
@@ -0,0 +1,120 @@
+/**
+ * @file
+ * @brief Stable Marriage Problem implementation
+ * @details
+ * This implementation utilizes the Gale-Shapley algorithm to find stable matches
+ * between men and women based on their preferences.
+ *
+ * **Stable Marriage Problem** aims to find a stable matching between two equally sized 
+ * sets of elements given an ordinal preference for each element. The algorithm was
+ * introduced by David Gale and Lloyd Shapley in 1962.
+ *
+ * @author [B Karthik](https://github.com/BKarthik7)
+ */
+
+#include <iostream>
+#include <vector>
+#include <cstring>
+#include <algorithm>
+using namespace std;
+
+void stableMatch(const vector<vector<int>>& women_prefs, const vector<vector<int>>& men_prefs, int n) {
+    vector<int> engagements(n, -1);
+    vector<bool> free_men(n, true);
+    vector<int> next_proposal(n, 0); // Tracks the next woman to propose for each man
+
+    while (true) {
+        int freeMan;
+        for (freeMan = 0; freeMan < n; freeMan++) {
+            if (free_men[freeMan]) break;
+        }
+        if (freeMan == n) break;
+
+        int womanToPropose = men_prefs[freeMan][next_proposal[freeMan]];
+        next_proposal[freeMan]++;
+        int currentFiance = engagements[womanToPropose];
+
+        if (currentFiance == -1) {
+            engagements[womanToPropose] = freeMan;
+            free_men[freeMan] = false;
+        } else {
+            if (find(women_prefs[womanToPropose].begin(), women_prefs[womanToPropose].end(), freeMan) <
+                find(women_prefs[womanToPropose].begin(), women_prefs[womanToPropose].end(), currentFiance)) {
+                engagements[womanToPropose] = freeMan;
+                free_men[freeMan] = false;
+                free_men[currentFiance] = true;
+            }
+        }
+    }
+
+    cout << "Stable Matches:\n";
+    for (int woman = 0; woman < n; woman++) {
+        cout << "Woman " << woman << " is engaged to Man " << engagements[woman] << endl;
+    }
+}
+
+void runTestCase(int caseNum) {
```

Comment:
```suggestion
void tests() {
```
please document all functions